### PR TITLE
[MLIR][LLVM] Add invariant intrinsics

### PR DESCRIPTION
**Pull request by gysit**
This commit implements the LLVM IR invariant intrinsics in LLVM dialect. These intrinsics can be used to mark a program regions in which the contents of a specific memory object will not change.

The LLVM dialect implementation also implements the PromotableOpInterface to ensure Mem2Reg & SROA are able to promote pointers that are marked using the invariant intrinsics.

### CODE REVIEW

**Comment by gysit** [comment]
Turns out the intrinsic returns an pointer that cannot have an address space. I will update...

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOpBase.td [hunk 1]
```diff
@@ -67,6 +67,9 @@ class LLVM_PointerInAddressSpace<int addressSpace> : Type<
     # addressSpace # ")";
 }
 
+// Type constraint accepting any LLVM pointer type in address space 0.
+def LLVM_Pointer : LLVM_PointerInAddressSpace<0>;
```
Comment (Dinistro) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMOpBase.td:71]: That seems a bit odd. Shouldn't this have a name that indicates that it's in the default address space?

File: mlir/test/Dialect/LLVMIR/sroa.mlir [hunk 2]
```diff
@@ -170,12 +170,14 @@ llvm.func @direct_promotable_use_is_fine_on_accessor() -> i32 {
...
   %1 = llvm.alloca %0 x !llvm.struct<"foo", (i32, f64, i32)> {alignment = 8 : i64} : (i32) -> !llvm.ptr
   %2 = llvm.getelementptr %1[0, 2] : (!llvm.ptr) -> !llvm.ptr, !llvm.struct<"foo", (i32, f64, i32)>
+  // This does not provide side-effect info but it can be removed because it implements PromotableOpInterface.
+  %3 = llvm.intr.invariant.start 2, %2 : !llvm.ptr
   // CHECK: %[[RES:.*]] = llvm.load %[[ALLOCA]]
-  %3 = llvm.load %2 : !llvm.ptr -> i32
+  %4 = llvm.load %2 : !llvm.ptr -> i32
   // This does not provide side-effect info but it can be removed because it implements PromotableOpInterface.
-  llvm.intr.lifetime.start 2, %2 : !llvm.ptr
+  llvm.intr.invariant.end %3, 2, %2 : !llvm.ptr
```
Comment (Dinistro) [inline mlir/test/Dialect/LLVMIR/sroa.mlir:178]: Is this related to this PR?

**Review by Dinistro — COMMENTED** [review]
Dropped a few comments.

File: mlir/test/Dialect/LLVMIR/sroa.mlir (hunk 2 above)
Comment (gysit) [reply mlir/test/Dialect/LLVMIR/sroa.mlir:178]: Yes I wanted to use the invariant intrinsic in one of the test cases. Above I there is still a test that exercises the lifetime intrinsic. Both of them implement the PromotableOpInterface that allows SROA to drop them if a memory slot is promoted.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOpBase.td (hunk 1 above)
Comment (gysit) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOpBase.td]: I went for LLVM_DefaultPointer now.

**Review by Dinistro — APPROVED** [review]
LGTM!
